README.md
MergePDF
========
MergePDF can merge many PDF/image files into one PDF file.

The main feature of MergePDF is that it determines the order of binding the PDF/image files according to their view sorting order in the Finder. The PDF files will merge with the order displayed.

Additionally, MergePDF can create bookmarks named according to their original file names.

## Usage
English :
* http://www.script-factory.net/software/misc/MergePDF/manual/en.lproj/index.html
Japanese :
* http://www.script-factory.net/software/misc/MergePDF/manual/ja.lproj/index.html

## Building
Reqirements :
* Mac OS X 10.11.4 or later.
* Xcode 7.3 or later.
* [ModuleLoader]

[ModuleLoader]: http://www.script-factory.net/software/OSAX/ModuleLoader/en/index.html

## Licence
Copyright &copy; 2007-2016 Tetsuro Kurita
Licensed under the [GPL license][GPL].
 
[GPL]: http://www.gnu.org/licenses/gpl.html



History.txt
MergePDF

== To Do
* イメージを PDF に変換するサービスメニューで、同名のPDFがある場合のエラー処理。
* ファイルダイアログを開くようにする。
* 白黒画像のときは、JBIG2 を使えないか。
  - PDFPage's initWithImage を使ったらどうにかならないか？
* eps ファイルを pdf ファイルに変換できるようにする。
  - ビットマップデータが含まれる時にちゃんと圧縮することができるか。
* ARC
* localize tool の更新

== 2016-02-02
* localize folder を en と ja にする。
  - project を閉じる。
  - git mv で English.lproj と Japanese.lproj を en.lproj と ja.lproj にする。
  - project を開き、en.lproj の中のファイルを project に add する。
  - Localization で japanese を選択。ダイアログで Use file を選ぶ。
* DropBox を導入して、ソースの場所を設定できるようにする。

== 2016-02-02
* HelpBook のフォルダレイアウトを更新する。
* modern Obj-C

== 2016-01-31
* runloop に戻らないと、すべてのNSTextField の値が変更されない。
  - [[NSRunLoop currentRunLoop] runMode:NSDefaultRunLoopMode beforeDate:[NSDate dateWithTimeIntervalSinceNow: 0.1]];
    を挿入する。
* PDFMerge.m を refactoring
* deprecated API を replace
* Sparkle の登録
* DonationReminder のサブプロジェクト化
* ライブラリスクリプトのサブモジュール化
* SmartActivateLib への依存を無くす。
* git に移行
* AppleEventExtra.h から NDCarbonUtiliteis への依存性を無くす
  - typeAlias から NSURL をえる。
  - typeFileURL に[NSAppleEventDescriptor coerceToDescriptorType:] 
  - [NSAppleEventDescriptor data] から、theURLValue = (NSURL *)CFBridgingRelease(CFURLCreateWithBytes(NULL, [urlData bytes], [urlData length], kCFStringEncodingUTF8, NULL));

== 2016-01-19
* AppleScriptKit 関係を削除する。
* XCode4.6 と 10.8 SDK に移行

== 2013-09-26
AppleScriptObjC に移行する。 -- 済み
* AppleScriptKit へのリンクをやめる。-- 済み
* nib から AppleScript Info の削除 -- 済み
* SM に登録する。-- 済み
* MergePDF(31472,0xb04b1000) malloc: reference count underflow for 0x1a25580, break on auto_refcount_underflow_error to debug.
  - GC に関連したメモリ管理周りに原因があるのかも
  - PDFMerger.m/(void)stat: のなかに [PDFDocument writeToFile:] を実行すると発生する。
  - PDFDocument のインスタンス化と writeToFile の間の処理を極力削除したが、上記のエラーは変わらず。
  - PDFDocument のバグと判断して、対応をあきらめる。
  - クラッシュすることはなさそう。
* 以下の手順でエラーが発生する。-- 済み
  - Progress window で OK ボタン
  - direction chooser window
  - save panel が表示され cancel
  - progress window で OK ボタン
  - エラー
  - list 表示にして、direction chooser が無くてもエラー

== 2013-09-25
* NSURL+NDCarbonUtilities の導入
  - URLWithFSRef を使えるようにする。
* ProgressWindow.nib 内に ProgressWindowController.applescript をインスタンス化 -- 済み
* ProgressWindow.nib の AppleScript コネクションを外す。 -- 済み
* MergePDF.applescript の内容は、AppController に移行する。 -- 済み
* メニュー/Open... の実装 -- 済み
* MainMenu.nib から AppleScript へのコネクションを外す。-- 済み
* main.m の変更 -- 済み
* ガーベージコレクションを有効にする。（必須にする）-- 済み
* Linked Framework に AppleScriptObjC.framework を追加。
* デプロイメントターゲット を Mac OS X 10.6 にする。 -- 済み
* SDK を Mac OS X 10.6 にする。 -- 済み

== 2011-04-15
* Convert Images to PDFs をとりあえず実装。

== 2011-04-14
なぜが、フォルダのドロップで起動すると、curerentEvent が null になる。
QuickFileType ではそんなことはない勝った。
DocumentBasedApplication じゃないから？
AppleScript Studio だから？

== 2011-04-08

PDFKit の　initwithImage ではなく、Core Graphics の PDFContext を使ったら、jpeg のファイルサイズは変わらない。
jpeg だけ、PDFContext で、ほかは initwithImage を使う。

Mac OS X 10.6 ではどうなるのか。
Mac OS X 10.6.6 では、initwithImage で jpeg ファイルはでっかくならない。Mac OS X 10.5 から、改良されいるのか？

以下解決
* jpeg をmerge するとサイズが爆発的に増える。
* 保存時に quartz filter を用いると、再圧縮できるらしい。
* merge するまえに、写真は jpeg ではなく あらかじめ PDF にしておけば大きくならない。


jpeg から PDF への変換の比較

* プレビュー : ファイルサイズ変化なし、DPI 変化 <-- Mac OS X 10.6 なら DPIは保存される。
* Automator : ファイルサイズ大、DPI 変化なし
* sips : ファイルサイズ変化なし、DPI 変化

Mac OS X 10.6 なら ファイルサイズ変化なし、DPI 保存。

== 2009-09-29
DonationWindow が開かれる時、次のようなエラーが発生している。 -- 済み
2009-09-29 09:03:54.021 MergePDF[24292:10b] *** -[DonationReminder sourceLocation]: unrecognized selector sent to instance 0x237ec0
このため、DonationWindow が開くと、MergePDF Window が開かれない。

== 2009-03-19
機能は全部 implement した
あとは、Sparkle のみ

== 2009-03-17
* pdf の作成はできるようになった。
To Do
* NSWindowController をつかって、複数の window を開けるようにする。
* WindowController に処理するフォルダを与える。
* AppleScript も object 単位でアサインする。
* daialog は全部 progress window のシートにする。
* だからとりあえず、progres window を表示する。
* AppController に、フォルダ参照を渡して progress window をひらく
* window controller が folder 参照を保持する。
* progress window が開いてから実作業がスタートする。

* progres window への状態の表示
* error message の表示
* progress window を floating でなくす
* icon view の merge 方向のデフォルトボタンをラベルの位置で決定する。
* Sparkle の組み込み


== To Do
* progress window

== 2008-01-22
以下済み
* image を UTI で判別できないか？info for の type identifier が使えるかも
		* file type が "PDF "
		* type identifier が "com.adobe.pdf"
		* name extension が pdf
		* public.jpeg
		* public.png
		* public.tiff
		* com.compuserve.gif
		* com.microsoft.bmp -- CGImageSourceCreateWithURL で読み込めない
		* com.apple.pict
* image を PDF に変換して取り込めるようにする。

== 2008.01.18
以下済み
* DonationReminder.framework を使う
* SmartActivate を活用する。
* FileSorter に横にlabel が並んでいる場合を考慮できるようにする。 
* SortEngine を Quick Sort にする。
* Localize は English を default にする。

== 2006.07.26
mergePDFto を導入したおかげで、動作しなくなっていたようだ。

== 2006.07.10
modal でない save panel は sheet にしないと表示できない。
Cocoa でも AppleScript Studio　でもかんけいない。

== 2006.07.07
ASKNibObjectInfo から script name を取得 -- これは難しそう
ASKScriptCache から script name で ASKScript を取得
ASKScript は OSAScript のサブクラス
script name はきめうちでよかろう。

* ファイル保存ダイアログが modal で表示される。mode less にならないものか？
	* Objective-C から AppleScript を呼ぶことができれば解決できる。
* donation 催促 window を追加。
* PDF を Acrobat で開いている場合、info for コマンドでファイルの busy status に反映されない。
lsof コマンドを使うようにする。

== 2004.12.22
png などの画像ファイルも扱えるようにしたかったが、insertPages で画像を扱えないようだ。

== 2004.09.03
1.6.3b1
日本語のbookmarkが化けることがある不具合を修正。
TEC OSAX ではなく、iconv を使うようにした。
PDFファイルを任意の場所に保存できない不具合を修正

必要なOSAX
TextInfo OSAX

== 2003.09.09
1.6.1b1
最新のFileSorter
TEC OSAXを不必要にした。
dittoをつかって2byte codeの名前のファイルをコピーするとAcrobatで開いた時おかしくなるようだ。Finderのdupulicateをつかいたい・・・名前のかえてのコピーができない。Acrobatで最初のfileをsave asする。
ShellUtilsがいらなくなった

必要なOSAX
TextInfo OSAX
TEC OSAX

== 2003.05.31
1.6b1
bookmarkの挿入はできるようになったが、挿入するpage数をカウントするところで、どうしても固まる。
昨日はそんなことはなかった。どうして？
mergeできるPDFは1pageの物だけに制限しよう。
mergeするpdfのpageすうはあらかじめ調べてからmergeするようにしたら、errorが回避できた。
2003.05.30
pageの挿入はjavascriptでなく、applescriptで実行できた
bookmarkの生成がどうしてもできない。
To Do
Marge PDFにBookmarkをつけるかどうかの設定を加える。

必要なOSAX
TextInfo OSAX
Akua Sweets

== 2003.01.08
1.5.4b1
ファイルダイアログでファイル名を入力しても、それが反映されない不具合を修正

== 2002.09.13
1.5.3b2
pathにdouble quoteとsingle quoteが両方入っているときは、諦めることにした。

PDFMerger.m
#import "PDFMerger.h"
#import "AppleEventExtra.h"

#define useLog 0

#define DEFAULT_DPI 72

@implementation PDFDestination (MergePDF)

+ (PDFDestination *)destinationWithPage:(PDFPage *)page
{
	NSRect rect = [page boundsForBox:kPDFDisplayBoxMediaBox];
	NSPoint point = NSMakePoint(kPDFDestinationUnspecifiedValue, rect.size.height);
	PDFDestination *dest = [[PDFDestination alloc] initWithPage:page atPoint:point];
	return dest;
}

@end

@implementation PDFDocument (MergePDF)
	
ImageKind image_type(NSString *path)
{
    NSWorkspace *ws = [NSWorkspace sharedWorkspace];
    NSError *err = nil;
    NSString *uti = [ws typeOfFile:path error:&err];
    if (err) {
        [NSApp presentError:err];
        return NotImage;
    }
    
    if ([ws type:uti conformsToType:@"com.adobe.pdf"]) {
        return PDFImage;
        
    } else if ([uti isEqualToString:@"com.adobe.illustrator.ai-image"]) {
        return PDFImage;
        
    } else if ([ws type:uti conformsToType:@"public.jpeg"]) {
        return JpegImage;
    } else if ([ws type:uti conformsToType:@"public.image"]) {
        return GenericImage;
    }
    
    return NotImage;
}

//deprecated use pdfDocumentWithImageURL:
// it looks PDFPage's initWithImage cause same result in almost
// But CGPDFContext help to keep file size of jpeg in Mac OS X 10.5
// In Mac OS X 10.6, initWithImage does not increse data size of jpeg.
+ (PDFDocument *)pdfDocumentWithImageFile:(NSString *)path 
{
	PDFDocument *doc = nil;
	CGContextRef out_context = NULL;
	size_t img_count;
	NSURL *url = [NSURL fileURLWithPath:path];
	CGImageSourceRef image_source =  CGImageSourceCreateWithURL((__bridge CFURLRef)url, NULL);
	CFMutableDataRef data = CFDataCreateMutable(NULL, 0);
	CGDataConsumerRef data_consumer = CGDataConsumerCreateWithCFData(data);
	if (data_consumer == NULL) {
		img_count = 0;
		goto bail;
	}
	out_context = CGPDFContextCreate(data_consumer, NULL, NULL);
	img_count = CGImageSourceGetCount(image_source);
	CFNumberRef dpi = NULL;
	for (size_t i = 0; i<img_count; i++) {
		CFDictionaryRef image_info = CGImageSourceCopyPropertiesAtIndex(image_source, i, NULL );
		CGImageRef image = CGImageSourceCreateImageAtIndex( image_source, i, NULL );
#if useLog
		 CFShow(image_info);
		 CFShow(CFDictionaryGetValue(image_info, kCGImagePropertyDPIHeight));
		 CFShow(CFDictionaryGetValue(image_info, kCGImagePropertyDPIWidth));
		 CFShow(CFDictionaryGetValue(image_info, kCGImagePropertyPixelHeight));
		 CFShow(CFDictionaryGetValue(image_info, kCGImagePropertyPixelWidth));
#endif
		float dpi_w = DEFAULT_DPI;
		dpi = CFDictionaryGetValue(image_info, kCGImagePropertyDPIWidth);
		if (dpi) {
			CFNumberGetValue(dpi, kCFNumberFloat32Type, &dpi_w);
		}
			
		float dpi_h = DEFAULT_DPI;
		dpi = CFDictionaryGetValue(image_info, kCGImagePropertyDPIHeight);
		if (dpi) {
			CFNumberGetValue(dpi, kCFNumberFloat32Type, &dpi_h);
		}
		
		int size_w;
		CFNumberGetValue(CFDictionaryGetValue(image_info, kCGImagePropertyPixelWidth), 
						 kCFNumberSInt32Type, &size_w);
		int size_h;
		CFNumberGetValue(CFDictionaryGetValue(image_info, kCGImagePropertyPixelHeight), 
						 kCFNumberSInt32Type, &size_h);
		CFRelease(image_info);
		
		CGRect img_rect = CGRectMake(0, 0, size_w/(dpi_w/DEFAULT_DPI), size_h/(dpi_h/DEFAULT_DPI));
		
		
		CGContextBeginPage(out_context, &img_rect);
		//CGPDFContextBeginPage(out_context, &img_rect);
		// Draw the image into the rect.
		CGContextDrawImage(out_context, img_rect, image);
		//CGContextEndPage(out_context);
		CGPDFContextEndPage(out_context);
		CFRelease(image);
		
	}
	CGPDFContextClose(out_context);
	CGContextRelease(out_context);
	if (CFDataGetLength(data)) {
		doc = [[PDFDocument alloc] initWithData:(NSData *)CFBridgingRelease(data)];
	}
bail:
	CFRelease(image_source);
	CGDataConsumerRelease(data_consumer);
	CFRelease(data);	
	return doc;	
}

//deprecated use pdfDocumentWithImageURL
+ (PDFDocument *)pdfDocumentWithPath:(NSString *)path
{
	PDFDocument *result = NULL;

	switch (image_type(path)) {
		case JpegImage:
			result = [PDFDocument pdfDocumentWithImageFile:path];
			break;
		case GenericImage:
        {
			result = [[PDFDocument alloc] init];
			NSImage *image = [[NSImage alloc] initWithContentsOfFile:path];
			NSEnumerator *enumerator = [[image representations] objectEnumerator];
			NSImageRep *imagerep;
			NSUInteger ind = 0;
			PDFPage *page = NULL;
			NSImage *single_image = NULL;
			while (imagerep = [enumerator nextObject]) { //support for multipage tiff
				single_image = [NSImage new];
				[single_image addRepresentation:imagerep];
				page = [[PDFPage alloc] initWithImage:single_image];
				[result insertPage:page atIndex:ind];
				ind++;
			}
			break;
        }
		default:
			result = [[PDFDocument alloc] initWithURL: [NSURL fileURLWithPath: path]];
			break;
	}
	return result;
}

+ (PDFDocument *)pdfDocumentWithImageURL:(NSURL *)fURL
{
    PDFDocument *result = [[PDFDocument alloc] init];
    NSImage *image = [[NSImage alloc] initWithContentsOfURL:fURL];
    NSArray *image_reps = [image representations];
    if ([image_reps count] > 1) {
        //support for multipage tiff
        NSUInteger ind = 0;
        for (NSImageRep *imgrep in image_reps) {
            NSImage *single_image = [NSImage new];
            [single_image addRepresentation:imgrep];
             PDFPage *page = [[PDFPage alloc] initWithImage:single_image];
             [result insertPage:page atIndex:ind++];
        }
    } else {
        PDFPage *page = [[PDFPage alloc] initWithImage:image];
        [result insertPage:page atIndex:0];
    }
    
    return result;
}

+ (PDFDocument *)pdfDocumentWithURL:(NSURL *)fURL
{
    PDFDocument *result = NULL;
    
	switch (image_type([fURL path])) {
		case JpegImage:
			result = [PDFDocument pdfDocumentWithImageURL:fURL];
			break;
		case GenericImage:
            result = [PDFDocument pdfDocumentWithImageURL:fURL];
			break;
		default:
			result = [[PDFDocument alloc] initWithURL: fURL];
			break;
	}
	return result;
}

- (void)appendOutline:(PDFOutline *)outline
{
	[[self outlineRoot] insertChild:outline atIndex:[[self outlineRoot] numberOfChildren]];	
}	
	
- (PDFOutline *) appendBookmark:(NSString *)label atPageIndex:(NSUInteger)index
{
	PDFOutline *newoutline = [[PDFOutline alloc] init];
	[newoutline setLabel:label];
#if useLog	
	NSLog(@"pageCount: %d, index:%d", [self pageCount], index);
#endif
	PDFPage *target_page = [self pageAtIndex:index];
	PDFDestination *bookmark_destination = [PDFDestination destinationWithPage:target_page];
	[newoutline setDestination:bookmark_destination];
	[self appendOutline:newoutline];	
	return newoutline;
}

// deprecated use mergeFileAtURL:
- (BOOL)mergeFile:(NSString *)path error:(NSError **)error
{
#if useLog
	NSLog(@"start mergeFile for %@", path);
#endif
	PDFDocument *pdf_doc = [PDFDocument pdfDocumentWithPath:path];
	if (!pdf_doc) {
		//NSLog(@"Fail to get PDF for %@", path);
		NSDictionary *dict = @{NSLocalizedDescriptionKey: [NSString stringWithFormat:@"Fail to get PDF for %@.",path]};
		*error = [NSError errorWithDomain:@"MergePDFErrorDomain" code:0 userInfo:dict];
		return NO;
	}
	NSInteger npages = [pdf_doc pageCount];
#if useLog
	NSLog(@"number of pages before appending : %d", [self pageCount]);
	NSLog(@"number of pages to append : %d", npages);
#endif
	for (int n = 0; n < npages; n++) {
		[self insertPage:[pdf_doc pageAtIndex:n] atIndex:[self pageCount]];
	}
#if useLog
	NSLog(@"number of pages after appending : %d", [self pageCount]);
#endif
	PDFOutline *outline = [pdf_doc outlineRoot];
	NSString *label = [[path lastPathComponent] stringByDeletingPathExtension];
	NSUInteger destpage_index = [self pageCount]-npages;
	if (outline) {
		PDFDestination *pdfdest = [PDFDestination destinationWithPage:[self pageAtIndex: destpage_index]];
		[outline setDestination:pdfdest];
		[outline setLabel:label];
		[self appendOutline:outline];
	} else {
		[self appendBookmark:[[path lastPathComponent] stringByDeletingPathExtension]
										   atPageIndex:destpage_index];
	}
	return YES;
}

- (BOOL)mergeFileAtURL:(NSURL *)fURL error:(NSError **)error
{
#if useLog
	NSLog(@"start mergeFile for %@", fURL);
#endif
	PDFDocument *pdf_doc = [PDFDocument pdfDocumentWithURL:fURL];
	if (!pdf_doc) {
		//NSLog(@"Fail to get PDF for %@", path);
		NSDictionary *dict = @{NSLocalizedDescriptionKey: [NSString stringWithFormat:@"Fail to get PDF for %@.",[fURL path]]};
		*error = [NSError errorWithDomain:@"MergePDFErrorDomain" code:0 userInfo:dict];
		return NO;
	}
	NSInteger npages = [pdf_doc pageCount];
#if useLog
	NSLog(@"number of pages before appending : %d", [self pageCount]);
	NSLog(@"number of pages to append : %d", npages);
#endif
	for (int n = 0; n < npages; n++) {
		[self insertPage:[pdf_doc pageAtIndex:n] atIndex:[self pageCount]];
	}
#if useLog
	NSLog(@"number of pages after appending : %d", [self pageCount]);
#endif
	PDFOutline *outline = [pdf_doc outlineRoot];
	NSString *label = [[fURL lastPathComponent] stringByDeletingPathExtension];
	NSUInteger destpage_index = [self pageCount]-npages;
	if (outline) {
		PDFDestination *pdfdest = [PDFDestination destinationWithPage:[self pageAtIndex: destpage_index]];
		[outline setDestination:pdfdest];
		[outline setLabel:label];
		[self appendOutline:outline];
	} else {
		[self appendBookmark:[[fURL lastPathComponent] stringByDeletingPathExtension]
                 atPageIndex:destpage_index];
	}
	return YES;
}

@end


@implementation PDFMerger

- (id)init {
    if (self = [super init]) {
		self.canceled = NO;
    }
    return self;
}

- (void)postProgressNotificationWithFile:(NSString *)path increment:(double)increment
{
	NSLog(@"postProgressNotificationWithFile:%@", path);
    NSNotificationCenter *noticenter = [NSNotificationCenter defaultCenter];
	NSString *message = [NSString stringWithFormat:NSLocalizedString(@"Processing %@", @""), 
							[path lastPathComponent]];
	NSDictionary *dict = @{@"message": message, 
							  @"levelIncrement": @(increment)};
	[noticenter postNotificationName:@"UpdateProgressMessage" object:self userInfo:dict];
}

- (void)postProgressNotificationWithMessage:(NSString *)message increment:(double)increment
{
	NSNotificationCenter *noticenter = [NSNotificationCenter defaultCenter];
	NSDictionary *dict = @{@"message": message, 
						  @"levelIncrement": @(increment)};
	//[noticenter postNotificationName:@"UpdateProgressMessage" object:self userInfo:dict];
    
    [noticenter performSelectorOnMainThread:@selector(postNotification:)
                                 withObject:[NSNotification notificationWithName:@"UpdateProgressMessage"
                                                                          object:self
                                                                        userInfo:dict]
                              waitUntilDone:NO];
}

- (void)postErrorNotification:(NSError *)error
{
	NSNotificationCenter *noticenter = [NSNotificationCenter defaultCenter];
	[noticenter postNotificationName:@"AppendErrorMessage" object:self userInfo:
		@{@"error": error}];
}

- (BOOL)checkCanceled
{
	if (!_canceled) {
#if useLog		
		NSLog(@"Not canceled");
#endif
		return NO;
	}
	[self postProgressNotificationWithMessage:@"Canceled" increment:0];
	return YES;
}

- (void)start:(id)sender
{
	@autoreleasepool {
        NSError *error = nil;
        if ([self checkCanceled]) return;
        double incstep = 85.0/[_targetFiles count];
        NSEnumerator *enumerator = [_targetFiles objectEnumerator];
        NSURL *fURL = [[enumerator nextObject] URL];
        [self postProgressNotificationWithFile:[fURL path] increment:incstep];
        PDFDocument *pdf_doc = [PDFDocument pdfDocumentWithURL:fURL];
        if (!pdf_doc) {
            NSDictionary *dict = @{NSLocalizedDescriptionKey: [NSString stringWithFormat:@"Fail to get PDF for %@.",[fURL path]]};
            error = [NSError errorWithDomain:@"MergePDFErrorDomain" code:0 userInfo:dict];
            [self postErrorNotification:error];
            return;
        }
        
        PDFOutline *outline = [pdf_doc outlineRoot];
        [pdf_doc setOutlineRoot:[[PDFOutline alloc] init]];
        NSString *label = [[fURL lastPathComponent] stringByDeletingPathExtension];
        if (outline) {
            [outline setLabel:label];
            PDFDestination *pdfdest = [PDFDestination destinationWithPage:[pdf_doc pageAtIndex:0]];
            [outline setDestination:pdfdest];
            [pdf_doc appendOutline:outline];	
        } else {
            [pdf_doc appendBookmark:label atPageIndex:0];
        }

        for (NSAppleEventDescriptor *aedesc in enumerator) {
            fURL = [aedesc URL];
            if ([self checkCanceled]) return;
            [self postProgressNotificationWithFile:[fURL path] increment:incstep];
            if (![pdf_doc mergeFileAtURL:fURL error:&error] ) {
                [self postErrorNotification:error];
            }
        }
        if ([self checkCanceled]) return;
        [self postProgressNotificationWithMessage:NSLocalizedString(@"Saving a new PDF file", @"") increment:5];
        [pdf_doc writeToFile:_destination];
        [self postProgressNotificationWithMessage:@"Success" increment:5];
    }

}

@end
